secure/json-rpc/src/lib.rs

```diff
@@ -0,0 +1,595 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The purpose of the JsonRpcClient presented here is to provide a lightweight and secure
+//! JSON RPC client to talk to the JSON RPC service offered by Libra Full Nodes. This is useful
+//! for various security-critical components (e.g., the secure key manager), as it allows
+//! interaction with the Libra blockchain in a minimal and secure manner.
+//!
+//! Note: While a JSON RPC client implementation already exists in the Libra codebase, this
+//! provides a simpler and (hopefully) more secure implementation with fewer dependencies.
+#![forbid(unsafe_code)]
+
+use hex::FromHexError;
+use libra_types::{
+    account_address::AccountAddress, account_state::AccountState,
+    account_state_blob::AccountStateBlob, transaction::SignedTransaction,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+use std::{convert::TryFrom, io};
+use thiserror::Error;
+use ureq::Response;
+
+/// Various constants for the JSON RPC client implementation
+const JSON_RPC_VERSION: &str = "2.0";
+const REQUEST_TIMEOUT: u64 = 10_000;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum Error {
+    #[error("JSON RPC call returned a custom internal error: {0}")]
+    InternalRPCError(String),
+    #[error("Data does not exist. Missing data: {0}")]
+    MissingData(String),
+    #[error("JSON RPC call failed with response: {0}")]
+    RPCFailure(String),
+    #[error("Serialization error: {0}")]
+    SerializationError(String),
+    #[error("Unknown error: {0}")]
+    UnknownError(String),
+}
+
+impl From<io::Error> for Error {
+    fn from(error: io::Error) -> Self {
+        Self::UnknownError(format!("{}", error))
+    }
+}
+
+impl From<lcs::Error> for Error {
+    fn from(error: lcs::Error) -> Self {
+        Self::SerializationError(format!("{}", error))
+    }
+}
+
+impl From<serde_json::Error> for Error {
+    fn from(error: serde_json::Error) -> Self {
+        Self::SerializationError(format!("{}", error))
+    }
+}
+
+impl From<FromHexError> for Error {
+    fn from(error: FromHexError) -> Self {
+        Self::SerializationError(format!("{}", error))
+    }
+}
+
+/// Provides a lightweight JsonRpcClient implementation.
+struct JsonRpcClient {
+    host: String,
+}
+
+// TODO(joshlind): remove this flag once the client is integrated into key manager. Otherwise,
+// rust complains this is never used...
+#[allow(dead_code)]
+impl JsonRpcClient {
+    pub fn new(host: &str, port: u16) -> Self {
+        let host = format!("http://{}:{}", host, port);
+        Self { host }
+    }
+
+    /// Submits a signed transaction to the Libra blockchain via the JSON RPC API.
+    pub fn submit_signed_transaction(
```

Comment:
I noticed today that the other API just has a "submit" function, :/ I like this more though